File: packages/client/test/day-night.test.ts
/**
 * The client half of the day/night cycle.
 *
 * Cross-language agreement is proven by golden vectors in `parity.test.ts`;
 * this file pins the shape of the function itself.
 */
import { describe, expect, it } from 'vitest';
import {
  DEFAULT_CYCLE_MS,
  sunAngle,
  sunElevation,
  timeOfDay,
} from '../src/world/day-night.js';

const CYCLE = 7_200_000;

describe('day/night', () => {
  it('completes exactly one turn per cycle', () => {
    expect(sunAngle(0, CYCLE)).toBe(0);
    expect(sunAngle(CYCLE, CYCLE)).toBeCloseTo(0, 9);
    expect(sunAngle(CYCLE / 2, CYCLE)).toBeCloseTo(Math.PI, 9);
  });

  it('is periodic, so a long absence needs no catching up', () => {
    for (let day = 0; day < 4; day++) {
      expect(sunAngle(day * CYCLE + 1000, CYCLE)).toBeCloseTo(sunAngle(1000, CYCLE), 9);
    }
  });

  it('peaks a quarter through and troughs three quarters through', () => {
    expect(sunElevation(CYCLE / 4, CYCLE)).toBeCloseTo(1, 9);
    expect(sunElevation((3 * CYCLE) / 4, CYCLE)).toBeCloseTo(-1, 9);
  });

  it('reports the fraction through the cycle', () => {
    expect(timeOfDay(CYCLE / 4, CYCLE)).toBeCloseTo(0.25, 9);
  });

  it('never leaves 0..2pi', () => {
    for (const t of [0, 1, CYCLE - 1, CYCLE * 17 + 3]) {
      const angle = sunAngle(t, CYCLE);
      expect(angle).toBeGreaterThanOrEqual(0);
      expect(angle).toBeLessThan(2 * Math.PI);
    }
  });

  it('falls back to the default cycle rather than dividing by zero', () => {
    expect(sunAngle(1000, 0)).toBe(sunAngle(1000, DEFAULT_CYCLE_MS));
  });
});

File: packages/server/test/world/day_night_test.exs
defmodule IwsdkPhoenix.World.DayNightTest do
  use ExUnit.Case, async: true

  alias IwsdkPhoenix.World.DayNight

  @cycle 7_200_000

  describe "sun_angle/2" do
    test "starts at zero and completes exactly one turn per cycle" do
      assert DayNight.sun_angle(0, @cycle) == 0.0
      assert_in_delta DayNight.sun_angle(@cycle, @cycle), 0.0, 1.0e-9
      assert_in_delta DayNight.sun_angle(div(@cycle, 2), @cycle), :math.pi(), 1.0e-9
    end

    test "is periodic — the same phase of any day gives the same angle" do
      # The property that makes it a pure function of time rather than state:
      # a client that reconnects after three days needs no catching up.
      for day <- 0..3 do
        assert_in_delta DayNight.sun_angle(day * @cycle + 1000, @cycle),
                        DayNight.sun_angle(1000, @cycle),
                        1.0e-9
      end
    end

    test "never leaves 0..2pi" do
      for t <- [0, 1, @cycle - 1, @cycle * 17 + 3] do
        angle = DayNight.sun_angle(t, @cycle)
        assert angle >= 0.0 and angle < 2 * :math.pi()
      end
    end
  end

  describe "sun_elevation/2" do
    test "peaks a quarter through the cycle and troughs three quarters through" do
      assert_in_delta DayNight.sun_elevation(div(@cycle, 4), @cycle), 1.0, 1.0e-9
      assert_in_delta DayNight.sun_elevation(3 * div(@cycle, 4), @cycle), -1.0, 1.0e-9
    end
  end

  describe "time_of_day/2" do
    test "is the fraction through the cycle" do
      assert DayNight.time_of_day(0, @cycle) == 0.0
      assert_in_delta DayNight.time_of_day(div(@cycle, 4), @cycle), 0.25, 1.0e-9
    end
  end

  test "a zero or negative cycle falls back to the default rather than dividing by zero" do
    assert DayNight.sun_angle(1000, 0) == DayNight.sun_angle(1000, DayNight.default_cycle_ms())
  end
end
